src/System.Slices/src/System/SpanEnumerators.cs

```diff
@@ -0,0 +1,106 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Collections;
+using System.Collections.Generic;
+
+namespace System
+{
+    public partial struct Span<T>
+    {
+        public Enumerator GetEnumerator()
+        {
+            return new Enumerator(this);
+        }
+
+        IEnumerator IEnumerable.GetEnumerator()
+        {
+            return new EnumeratorObject(this);
+        }
+
+        IEnumerator<T> IEnumerable<T>.GetEnumerator()
+        {
+            return new EnumeratorObject(this);
+        }
+
+        /// <summary>
+        /// A struct-based enumerator, to make fast enumerations possible.
+        /// This isn't designed for direct use, instead see GetEnumerator.
+        /// </summary>
+        public struct Enumerator
+        {
```

Comment:
The issue is mostly that once C# sees IDisposable (which unavoidably comes with IEnumerator), it will have to wrap the whole thing into "using". 
It impacts performance of foreach enough that it makes sense to have a separate light-weight struct enumerator just for foreach pattern.

Then, once you have two iterators, the one that implements IEnumerator is better be a class because you have to allocate anyways (new or box) and interface dispatch on boxed structs is noticeable slower.

We came to exactly same conclusions when implementing ImmutableArray.Enumerator
https://github.com/dotnet/corefx/blob/master/src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray%601%2BEnumerator.cs